File: 1706-where-will-the-ball-fall/1706-where-will-the-ball-fall.cpp
class Solution {
public:
    vector<int> findBall(vector<vector<int>>& grid) {
        vector<int> res;
        for (int i = 0; i < grid[0].size(); ++i) {
            res.push_back(findBallCol(grid, 0, i));
        }
        return res;
    }
private:
    int findBallCol(vector<vector<int>>& grid, int r, int c) {
        int m = grid.size(), n = grid[0].size();
        if (r == m) return c;
        int cNew = c + grid[r][c];
        if (cNew < 0 || cNew >= n || grid[r][c] != grid[r][cNew]) {
            return -1;
        }
        return findBallCol(grid, r + 1, cNew);
    }
};

File: 1706-where-will-the-ball-fall/1706-where-will-the-ball-fall.py
class Solution:
    def findBall(self, grid: List[List[int]]) -> List[int]:
        m, n = len(grid), len(grid[0])
        def dfs(r, c):
            if r == m: return c
            c_new = c + grid[r][c]
            if c_new < 0 or c_new > n - 1 or grid[r][c] != grid[r][c_new]:
                return -1
            return dfs(r + 1, c_new)
        
        res = []
        for i in range(n):
            res.append(dfs(0, i))
        return res